### PR TITLE
Brain: Move to PIOS_HAL

### PR DESCRIPTION
**Pull request by jihlein**
Implement PIOS_HAL on Brain.  Compiles clean, needs testing by someone with Brain FC board.

[Review on Reviewable](https://reviewable.io/reviews/d-ronin/dronin/411)

### CODE REVIEW

**Comment by mlyle** [comment]
This is awesome!  Bumping to next release, though.

File: shared/uavobjectdefinition/hwbrain.xml [hunk 1]
```diff
@@ -2,67 +2,80 @@
...
 			</options>
 		</field>
 
-		<field name="FlxPort" units="function" type="enum" elements="1"  defaultvalue="Disabled">
+		<field name="FlxPort" units="function" type="enum" elements="1" parent="HwShared.PortTypes" defaultvalue="Disabled">
 			<options>
 				<option>Disabled</option>
-                <option>I2C</option>
-				<option>GPS</option>
+                <option>ComBridge</option>
```
Comment (fujin) [inline shared/uavobjectdefinition/hwbrain.xml]: indent

**Comment by fujin** [comment]
freakin' sweet, been wanting to do this for ages!

**Comment by mluessi** [comment]
Very nice @jihlein! I agree with @mlyle, it's safer to postpone merging this until after the release.

**Comment by mlyle** [comment]
Just a test of jenkins

jenkins, build artifacts please

**Comment by mlyle** [comment]
_[shared/uavobjectdefinition/hwbrain.xml, line 86 [r2]](https://reviewable.io:443/reviews/d-ronin/dronin/411#-K8XLainyMox_Cz5cOWh:-K8XLainyMox_Cz5cOWi:-1175131850) ([raw file](https://github.com/d-ronin/dronin/blob/1cbcb5a29ca4d61e544a3a776a848a3f03d9764b/shared/uavobjectdefinition/hwbrain.xml#L86)):_
These should not change (indent is now wrong).

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/411)_

**Comment by mlyle** [comment]
Overall looks good to me other than the whitespace change.  But this one should actually be thoroughly tested before merge.

File: flight/targets/brain/fw/pios_board.c [hunk 2]
```diff
@@ -882,54 +484,81 @@ void PIOS_Board_Init(void) {
...
-		PIOS_Board_configure_com(&pios_rxportusart_cfg, 0, PIOS_COM_FRSKYSENSORHUB_TX_BUF_LEN, &pios_usart_com_driver, &pios_com_frsky_sensor_hub_id);
-#endif /* PIOS_INCLUDE_FRSKY_SENSOR_HUB */
+		// special mode that enables PPM, FrSky RSSI, and Sensor Hub
+		PIOS_HAL_ConfigurePort(HWSHARED_PORTTYPES_FRSKYSENSORHUB,  // port type protocol
+				NULL,                                              // usart_port_cfg
+				&pios_rxportusart_cfg,                             // frsky usart_port_cfg
+				NULL,                                              // com_driver
+				NULL,                                              // i2c_id
+				NULL,                                              // i2c_cfg
+				NULL,                                              // ppm_cfg
```
Comment (mluessi) [inline flight/targets/brain/fw/pios_board.c:508]: PPM should be set in this mode

**Comment by mluessi** [comment]
Other than the PPM thing, this looks good to me.

**Comment by jihlein** [comment]
The case statement on line 506 is not terminated with a break, so rather than breaking out of the switch list shouldn't it fall thru to the PPM setup below?

**Comment by ufoDziner** [comment]
jenkins, build artifacts please

**Comment by ufoDziner** [comment]
S.Bus did not work.  Battery module showed correct voltage.  Couldn't test arming or flight due to the S.Bus issue.  Config image attached.
brain

**Comment by jihlein** [comment]
Sbus toggle for the main port is set to false, so the inverter is not getting set. I'll push a fix in the morning.

**Comment by mlyle** [comment]
@jihlein   can you please double check the other ones we merged when you're able?

**Comment by jihlein** [comment]
Push fix for Brain target.

Colibri looks okay, FlyingF3 and DiscoveryF4 don't use the field in question.  I believe we are okay.

**Comment by mlyle** [comment]
At least for sbus on mainport seems to work.  Can anyone else test any advanced functionality?

**Comment by mlyle** [comment]
Flown today with sbus.

**Comment by droidicus** [comment]
Tested along with #520, #411, #563, #139 on BrainFPV.

Bench test only, no flight. Was able to connect to Brain and modify settings, etc. FC is currently a bare board, so no additional tests were performed.

**Comment by pug398** [comment]
jenkins, build artifacts please

**Comment by pug398** [comment]
Apparently I have some glitches on my sbus as I gets lots of false triggering during radio wizard setup. Since no one else has reported this I will swap out sbus rx and hope to give Brain AH, PH, RTH and Path Following a try shortly.

**Comment by mlyle** [comment]
@pug398 -- We do have #68 open for flaky false triggering in radio wizard.  @Varcht was actually looking at working around that today.

**Comment by pug398** [comment]
X8R seems much better but I did get an initial false trigger on the throttle channel. Telemetry on rx port uart seems to not work where it did before. Not sure if that is related to this PR or not.

**Comment by jihlein** [comment]
@pug398, which type of telemetry?

**Comment by pug398** [comment]
Native telemetry. Rx Port set to UART. RX Port UART set to Telemetry (57k).
Was working on Renatus with a serial data radio but I may have broken something.

**Comment by jihlein** [comment]
Regarding the RxPort, I see an error that will prevent FrSky Sensor Hub Telemetry from working in the PPM+FrSky mode, but I think standard telemetry should be okay.

I'll push the PPM+FrSky fix this evening.

**Comment by pug398** [comment]
Thanks for looking it over. I will put renatus back on and see if it starts working again.

**Comment by pug398** [comment]
Ok telemetry on rx port uart is definitely working on Renatus3 so there is probably an issue with PR or next. I will see if I can determine which.

**Comment by jihlein** [comment]
@pug398, it's mostly like with the PR.  I'm reviewing too.

**Comment by pug398** [comment]
next works ok...

**Comment by mlyle** [comment]
@jihlein -- my sympathies, i don't see anything obviously wrong with this.

**Comment by pug398** [comment]
How likely firmware corrupted? I have been using the same one.

**Comment by mlyle** [comment]
Not likely-- the build system is consistent/trustworthy.

**Comment by mlyle** [comment]
@jihlein I found it-- looks like you lost the case HWBRAIN_RXPORT_UART:   label in the case statement.

**Comment by jihlein** [comment]
@mlyle, thanks!  I must have read over those lines a dozen times.

@pug398, please give it a try again when you have chance.

**Comment by mlyle** [comment]
jenkins, build artifacts please

**Comment by pug398** [comment]
@jihlein @mlyle Nice job that fixed the telemetry issue. Flew basic flight with no issues with a function on every port +OSD and everything looked good. Will turn on autonomous modes next daylight flight.

**Comment by mlyle** [comment]
@pug398 -- much appreciated, thank you for the test!

**Comment by mlyle** [comment]
:lgtm_strong: @jihlein--- thanks for your patience in getting this done.

---

Reviewed 1 of 4 files at r1, 1 of 1 files at r3, 1 of 2 files at r4, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/411#-:-K9ZPa02uTebTAhxrJR3:-1863009634)_

**Pull request merged**